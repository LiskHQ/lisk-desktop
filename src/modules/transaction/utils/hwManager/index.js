import { updateTransactionSignatures } from '@wallet/utils/hwManager';
import { signTransaction } from '@libs/hwManager/communication';
/**
 * signTransactionByHW - Function.
 * This function is used for sign a send hardware wallet transaction.
 */
const signTransactionByHW = async (
  wallet,
  networkIdentifier,
  transactionObject,
  transactionBytes,
  keys
) => {
  const data = {
    deviceId: wallet.hwInfo.deviceId,
    index: wallet.hwInfo.derivationIndex,
    networkIdentifier,
    transactionBytes,
  };

  try {
    const signature = await signTransaction(data);
    return updateTransactionSignatures(wallet, transactionObject, signature, keys);
  } catch (error) {
    throw new Error(error);
  }
};

// eslint-disable-next-line import/prefer-default-export
export { signTransactionByHW };
