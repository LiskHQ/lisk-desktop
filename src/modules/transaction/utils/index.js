export { default as normalizeStakesForTx } from './normalizeStakesForTx';
export * from './moduleCommand';
export * from './transaction';
export * from './transactionDetailsHelper';
export * from './removeDuplicateTransactions';
