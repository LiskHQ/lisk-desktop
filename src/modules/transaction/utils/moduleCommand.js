import { MODULE_COMMANDS_NAME_MAP } from 'src/modules/transaction/configuration/moduleCommand';

const getModuleCommandSenderLabel = (t = str => str) => ({
  [MODULE_COMMANDS_NAME_MAP.transfer]: t('Sender'),
  [MODULE_COMMANDS_NAME_MAP.reclaim]: t('Sender'),
  [MODULE_COMMANDS_NAME_MAP.unlock]: t('Sender'),
  [MODULE_COMMANDS_NAME_MAP.voteDelegate]: t('Voter'),
  [MODULE_COMMANDS_NAME_MAP.registerDelegate]: t('Account username'),
  [MODULE_COMMANDS_NAME_MAP.registerMultisignature]: t('Registrant'),
  [MODULE_COMMANDS_NAME_MAP.reportDelegateMisbehavior]: t('Reporter'),
});

const getModuleCommandTitle = (t = str => str) => ({
  [MODULE_COMMANDS_NAME_MAP.transfer]: t('Send'),
  [MODULE_COMMANDS_NAME_MAP.reclaim]: t('Reclaim'),
  [MODULE_COMMANDS_NAME_MAP.unlock]: t('Unlock'),
  [MODULE_COMMANDS_NAME_MAP.voteDelegate]: t('Vote'),
  [MODULE_COMMANDS_NAME_MAP.registerDelegate]: t('Register delegate'),
  [MODULE_COMMANDS_NAME_MAP.registerMultisignature]: t('Register multisignature'),
  [MODULE_COMMANDS_NAME_MAP.reportDelegateMisbehavior]: t('Report delegate misbehaviour'),
});

const splitModuleAndCommand = (moduleCommand) => moduleCommand.split(':');

const joinModuleAndCommand = ({ module, command }) => [module, command].join(':');

export {
  getModuleCommandTitle,
  joinModuleAndCommand,
  splitModuleAndCommand,
  getModuleCommandSenderLabel,
};
