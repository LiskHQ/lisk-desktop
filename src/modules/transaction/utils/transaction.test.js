/* eslint-disable max-lines */
import { cryptography } from '@liskhq/lisk-client';
import { MODULE_COMMANDS_NAME_MAP } from 'src/modules/transaction/configuration/moduleCommand';
import { splitModuleAndCommand } from 'src/modules/transaction/utils/moduleCommand';
import {
  getAddressFromBase32Address,
  getBase32AddressFromAddress,
} from '@wallet/utils/account';
import accounts from '@tests/constants/wallets';
import { genKey, blsKey, pop } from '@tests/constants/keys';
import {
  getTxAmount,
  elementTxToDesktopTx,
  containsTransactionType,
  desktopTxToElementsTx,
  transactionToJSON,
  removeExcessSignatures,
  convertStringToBinary,
  normalizeTransactionsStatisticsParams,
  normalizeNumberRange,
} from './transaction';

const address = 'lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt';
jest.spyOn(cryptography.address, 'getLisk32AddressFromPublicKey').mockReturnValue(address);

const {
  transfer, voteDelegate, registerMultisignature, registerDelegate, reclaim, unlock,
} = MODULE_COMMANDS_NAME_MAP;

// TODO: All of these tests need to be rewritten to adopt to new transaction schema https://github.com/LiskHQ/lisk-sdk/blob/7e71617d281649a6942434f729a815870aac2394/elements/lisk-transactions/src/schema.ts#L15
// We need to avoid lot of back and forth convertion from JSON and JS object
// For consistency we will adopt these changes similar to https://github.com/LiskHQ/lisk-sdk/blob/development/elements/lisk-api-client/src/transaction.ts
// We will address of these problem in issue https://github.com/LiskHQ/lisk-desktop/issues/4400

describe('API: LSK Transactions', () => {
  const baseDesktopTx = {
    sender: {
      publicKey: accounts.genesis.summary.publicKey,
      address: accounts.genesis.summary.address,
    },
    nonce: accounts.genesis.sequence.nonce,
    fee: '1000000',
    signatures: [],
  };
  const baseElementsTx = {
    senderPublicKey: convertStringToBinary(accounts.genesis.summary.publicKey),
    nonce: BigInt(accounts.genesis.sequence.nonce),
    fee: BigInt(1000000),
    signatures: [],
  };
  describe('getTxAmount', () => {
    it('should return amount of transfer in Beddows', () => {
      const tx = {
        moduleCommand: transfer,
        params: { amount: 100000000 },
      };

      expect(getTxAmount(tx)).toEqual(tx.params.amount);
    });

    it('should return amount of votes in Beddows', () => {
      const tx = {
        title: voteDelegate,
        moduleCommand: voteDelegate,
        params: {
          votes: [
            {
              amount: '100000000',
            },
            {
              amount: '100000000',
            },
          ],
        },
      };

      expect(getTxAmount(tx)).toEqual(200000000);
    });

    it('should return amount of unlock in Beddows', () => {
      const tx = {
        title: unlock,
        moduleCommand: unlock,
        params: {
          unlockObjects: [
            {
              amount: '100000000',
            },
            {
              amount: '100000000',
            },
          ],
        },
      };

      expect(getTxAmount(tx)).toEqual(200000000);
    });
  });

  describe('desktopTxToElementsTx', () => {
    it('creates a transaction object for transfer transaction', () => {
      const tx = {
        ...baseDesktopTx,
        moduleCommand: transfer,
        params: {
          recipient: { address: accounts.delegate.summary.address },
          amount: 100000000,
          data: 'test',
        },
      };
      const txObj = desktopTxToElementsTx(tx, transfer);
      const [module, command] = splitModuleAndCommand(transfer);
      expect(txObj).toEqual({
        ...baseElementsTx,
        module,
        command,
        params: {
          recipientAddress: expect.arrayContaining([]),
          amount: BigInt(100000000),
          data: 'test',
        },
      });
    });

    it('creates a transaction object for vote transaction', () => {
      const tx = {
        ...baseDesktopTx,
        moduleCommand: voteDelegate,
        params: {
          votes: [
            {
              amount: '100',
              delegateAddress: accounts.genesis.summary.address,
            },
            {
              amount: '-100',
              delegateAddress: accounts.delegate.summary.address,
            },
          ],
        },
      };
      const txObj = desktopTxToElementsTx(tx, voteDelegate);
      const [module, command] = splitModuleAndCommand(voteDelegate);
      expect(txObj).toEqual({
        ...baseElementsTx,
        module,
        command,
        params: {
          votes: tx.params.votes.map(item => ({
            amount: BigInt(item.amount),
            delegateAddress: expect.arrayContaining([]),
          })),
        },
      });
    });

    it('creates a transaction object for delegate registration transaction', () => {
      const tx = {
        ...baseDesktopTx,
        moduleCommand: registerDelegate,
        params: {
          username: 'username',
          generatorKey: genKey,
          blsKey: blsKey,
          proofOfPossession: pop,
        },
      };
      const txObj = desktopTxToElementsTx(tx, registerDelegate);
      const [module, command] = splitModuleAndCommand(registerDelegate);
      expect(txObj).toEqual({
        ...baseElementsTx,
        module,
        command,
        params: {
          username: 'username',
          generatorKey: convertStringToBinary(genKey),
          blsKey: convertStringToBinary(blsKey),
          proofOfPossession: convertStringToBinary(pop),
        },
      });
    });

    it('creates a transaction object for reclaimLSK transaction', () => {
      const tx = {
        ...baseDesktopTx,
        moduleCommand: reclaim,
        params: {
          amount: '10000000',
        },
      };
      const txObj = desktopTxToElementsTx(tx, reclaim);
      const [module, command] = splitModuleAndCommand(reclaim);
      expect(txObj).toEqual({
        ...baseElementsTx,
        module,
        command,
        params: {
          amount: BigInt('10000000'),
        },
      });
    });

    it('creates a transaction object for unlockToken transaction', () => {
      const unlockObjects = [
        { delegateAddress: accounts.genesis.summary.address, amount: '-1000' },
        { delegateAddress: accounts.delegate.summary.address, amount: '1000' },
      ];
      const tx = {
        ...baseDesktopTx,
        moduleCommand: unlock,
        params: {
          unlockObjects,
        },
      };
      const txObj = desktopTxToElementsTx(tx, unlock);
      const [module, command] = splitModuleAndCommand(unlock);
      expect(txObj).toEqual({
        ...baseElementsTx,
        module,
        command,
        params: {
          unlockObjects: tx.params.unlockObjects.map(item => ({
            amount: BigInt(item.amount),
            delegateAddress: expect.arrayContaining([]),
          })),
        },
      });
    });

    it('creates a transaction object for registerMultisignature transaction', () => {
      const tx = {
        ...baseDesktopTx,
        moduleCommand: registerMultisignature,
        params: {
          numberOfSignatures: 2,
          mandatoryKeys: [accounts.genesis.summary.publicKey, accounts.delegate.summary.publicKey],
          optionalKeys: [accounts.delegate_candidate.summary.publicKey],
        },
      };
      const txObj = desktopTxToElementsTx(tx, registerMultisignature);
      const [module, command] = splitModuleAndCommand(registerMultisignature);
      expect(txObj).toEqual({
        ...baseElementsTx,
        module,
        command,
        params: {
          numberOfSignatures: 2,
          mandatoryKeys: tx.params.mandatoryKeys.map(() => expect.arrayContaining([])),
          optionalKeys: tx.params.optionalKeys.map(() => expect.arrayContaining([])),
        },
      });
    });
  });

  describe('elementTxToDesktopTx', () => {
    it('should a transfer transaction with type signature of lisk service', () => {
      const [module, command] = splitModuleAndCommand(transfer);
      const tx = {
        ...baseElementsTx,
        module,
        command,
        params: {
          amount: BigInt(100000000),
          recipientAddress: getAddressFromBase32Address(accounts.delegate.summary.address),
          data: '',
        },
      };

      expect(elementTxToDesktopTx(tx)).toEqual({
        ...baseDesktopTx,
        moduleCommand: transfer,
        id: '',
        params: {
          amount: '100000000',
          recipient: { address: accounts.delegate.summary.address },
          data: '',
        },
      });
    });

    it('should a register delegate transaction with type signature of lisk service', () => {
      const [module, command] = splitModuleAndCommand(registerDelegate);
      const tx = {
        ...baseElementsTx,
        module,
        command,
        params: {
          username: 'super_delegate',
          generatorKey: convertStringToBinary(genKey),
          blsKey: convertStringToBinary(blsKey),
          proofOfPossession: convertStringToBinary(pop),
        },
      };

      expect(elementTxToDesktopTx(tx)).toEqual({
        ...baseDesktopTx,
        moduleCommand: registerDelegate,
        id: '',
        params: {
          username: 'super_delegate',
          generatorKey: genKey,
          blsKey,
          proofOfPossession: pop,
        },
      });
    });

    it('should a vote delegate transaction with type signature of lisk service', () => {
      const [module, command] = splitModuleAndCommand(voteDelegate);
      const tx = {
        ...baseElementsTx,
        module,
        command,
        params: {
          votes: [
            {
              amount: BigInt('100'),
              delegateAddress: getAddressFromBase32Address(accounts.delegate.summary.address),
            },
          ],
        },
      };

      expect(elementTxToDesktopTx(tx)).toEqual({
        ...baseDesktopTx,
        moduleCommand: voteDelegate,
        id: '',
        params: {
          votes: [
            {
              amount: '100',
              delegateAddress: accounts.delegate.summary.address,
            },
          ],
        },
      });
    });

    it('should transform a reclaimLSK transaction', () => {
      const [module, command] = splitModuleAndCommand(reclaim);
      const tx = {
        ...baseElementsTx,
        module,
        command,
        params: {
          amount: BigInt(100),
        },
      };

      expect(elementTxToDesktopTx(tx)).toEqual({
        ...baseDesktopTx,
        moduleCommand: reclaim,
        id: '',
        params: {
          amount: '100',
        },
      });
    });

    it('should transform a unlockToken transaction', () => {
      const [module, command] = splitModuleAndCommand(unlock);
      const unlockObjects = [
        {
          delegateAddress:
            getAddressFromBase32Address(accounts.delegate.summary.address),
          amount: BigInt('10000000'),
          unvoteHeight: 1000000,
        },
        {
          delegateAddress:
            getAddressFromBase32Address(accounts.send_all_wallet.summary.address),
          amount: BigInt('-10000000'),
          unvoteHeight: 1000000,
        },
      ];

      const tx = {
        ...baseElementsTx,
        module,
        command,
        params: { unlockObjects },
      };

      expect(elementTxToDesktopTx(tx)).toEqual({
        ...baseDesktopTx,
        moduleCommand: unlock,
        id: '',
        params: {
          unlockObjects: tx.params.unlockObjects.map(item => ({
            amount: String(item.amount),
            delegateAddress: getBase32AddressFromAddress(item.delegateAddress),
            unvoteHeight: item.unvoteHeight,
          })),
        },
      });
    });

    it('should transform a registerMultisignature transaction', () => {
      const [module, command] = splitModuleAndCommand(registerMultisignature);
      const mandatoryKeys = [
        accounts.genesis.summary.publicKey,
        accounts.delegate.summary.publicKey,
      ].map(key => convertStringToBinary(key));
      const optionalKeys = [
        accounts.delegate_candidate.summary.publicKey,
      ].map(key => convertStringToBinary(key));

      const tx = {
        ...baseElementsTx,
        module,
        command,
        params: {
          numberOfSignatures: 2,
          mandatoryKeys,
          optionalKeys,
        },
      };

      expect(elementTxToDesktopTx(tx)).toEqual({
        ...baseDesktopTx,
        moduleCommand: registerMultisignature,
        id: '',
        params: {
          numberOfSignatures: 2,
          mandatoryKeys: [
            accounts.genesis.summary.publicKey,
            accounts.delegate.summary.publicKey,
          ],
          optionalKeys: [
            accounts.delegate_candidate.summary.publicKey,
          ],
        },
      });
    });
  });

  describe('containsTransactionType', () => {
    it('should return true', () => {
      let pending = [{ moduleCommand: voteDelegate }];
      expect(containsTransactionType(pending, voteDelegate)).toEqual(true);

      pending = [{ moduleCommand: transfer }, { moduleCommand: voteDelegate }];
      expect(containsTransactionType(pending, voteDelegate)).toEqual(true);
    });

    it('should return false', () => {
      let pending = [];
      expect(containsTransactionType(pending, voteDelegate)).toEqual(false);

      pending = [{ moduleCommand: transfer }];
      expect(containsTransactionType(pending, voteDelegate)).toEqual(false);
    });
  });

  describe('transactionToJSON', () => {
    beforeEach(() => {
      // eslint-disable-next-line no-extend-native
      BigInt.prototype.toJSON = function () { return `${this.toString()}n`; };
    });

    afterEach(() => {
      // eslint-disable-next-line no-extend-native
      BigInt.prototype.toJSON = undefined;
    });
    const transaction = {
      ...baseElementsTx,
      module: 2,
      command: 0,
      params: {
        amount: BigInt(10000),
        recipientAddress: getBase32AddressFromAddress(accounts.delegate.summary.address),
        data: '',
      },
    };
    it('should return the transaction as JSON', () => {
      expect(JSON.parse(transactionToJSON(transaction)))
        .toEqual({
          senderPublicKey: accounts.genesis.summary.publicKey,
          nonce: '1n',
          fee: '1000000n',
          signatures: [],
          module: 2,
          command: 0,
          params: {
            amount: '10000n',
            recipientAddress: expect.stringContaining('lsk'),
            data: '',
          },
        });
    });
  });

  describe('removeExcessSignatures', () => {
    it('should remove optional signature considering the sender signature', () => {
      const nonEmpty = convertStringToBinary(accounts.genesis.summary.publicKey);
      const empty = convertStringToBinary('');
      const mandatoryKeysNo = 2;
      const hasSenderSignature = true;
      const signatures = [nonEmpty, nonEmpty, empty, nonEmpty];
      const expectSignatures = [nonEmpty, nonEmpty, empty, empty];
      expect(
        removeExcessSignatures(signatures, mandatoryKeysNo, hasSenderSignature),
      ).toEqual(expectSignatures);
    });
  });

  describe('normalizeTransactionsStatisticsParams', () => {
    it('returns expected API params', () => {
      let expectedResult = { interval: 'day', limit: 7 };
      expect(normalizeTransactionsStatisticsParams('week')).toEqual(expectedResult);
      expectedResult = { interval: 'month', limit: 6 };
      expect(normalizeTransactionsStatisticsParams('month')).toEqual(expectedResult);
      expectedResult = { interval: 'month', limit: 12 };
      expect(normalizeTransactionsStatisticsParams('year')).toEqual(expectedResult);
    });
  });

  describe('normalizeNumberRange', () => {
    it('returns distribution data by amount', () => {
      const distributionData = {
        "1_10": 1705,
        "10_100": 371,
        "100_1000": 287,
        "1000_10000": 180,
        "10000_100000": 53,
        "100000_1000000": 5,
        "0.1_1": 2258,
        "0.01_0.1": 697,
        "0.001_0.01": 306
      }
      const expectedResult = {
        "0 - 10 LSK": 4966,
        "11 - 100 LSK": 371,
        "101 - 1000 LSK": 287,
        "1001 - 10,000 LSK": 180,
        "10,001 - 100,000 LSK": 53,
        "100,001 - 1,000,000 LSK": 5
      }
      expect(normalizeNumberRange(distributionData)).toEqual(expectedResult);
    });
  });
});
