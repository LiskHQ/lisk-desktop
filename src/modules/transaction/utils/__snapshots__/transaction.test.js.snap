// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API: LSK Transactions createTransactionObject creates a transaction object for delegate registration transaction 1`] = `
Object {
  "asset": Object {
    "username": "username",
  },
  "assetID": 0,
  "fee": 1000000n,
  "moduleID": 5,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`API: LSK Transactions createTransactionObject creates a transaction object for reclaimLSK transaction 1`] = `
Object {
  "asset": Object {
    "amount": 10000000n,
  },
  "assetID": 0,
  "fee": 1000000n,
  "moduleID": 1000,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`API: LSK Transactions createTransactionObject creates a transaction object for registerMultisignatureGroup transaction 1`] = `
Object {
  "asset": Object {
    "mandatoryKeys": Array [
      Object {
        "data": Array [
          15,
          233,
          163,
          241,
          162,
          27,
          85,
          48,
          242,
          127,
          135,
          164,
          20,
          181,
          73,
          231,
          154,
          148,
          11,
          242,
          79,
          223,
          43,
          47,
          5,
          231,
          242,
          42,
          238,
          236,
          200,
          106,
        ],
        "type": "Buffer",
      },
      Object {
        "data": Array [
          134,
          73,
          152,
          121,
          68,
          141,
          27,
          2,
          21,
          213,
          156,
          191,
          7,
          136,
          54,
          227,
          215,
          217,
          210,
          120,
          45,
          86,
          162,
          39,
          74,
          86,
          135,
          97,
          191,
          243,
          111,
          25,
        ],
        "type": "Buffer",
      },
    ],
    "numberOfSignatures": 2,
    "optionalKeys": Array [
      Object {
        "data": Array [
          53,
          198,
          178,
          85,
          32,
          252,
          134,
          139,
          86,
          200,
          63,
          237,
          110,
          28,
          137,
          187,
          53,
          15,
          183,
          153,
          74,
          93,
          160,
          188,
          234,
          122,
          79,
          98,
          31,
          148,
          140,
          127,
        ],
        "type": "Buffer",
      },
    ],
  },
  "assetID": 0,
  "fee": 1000000n,
  "moduleID": 4,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`API: LSK Transactions createTransactionObject creates a transaction object for transfer transaction 1`] = `
Object {
  "asset": Object {
    "amount": 1n,
    "data": "test",
    "recipientAddress": Object {
      "data": Array [],
      "type": "Buffer",
    },
  },
  "assetID": 0,
  "fee": 1000000n,
  "moduleID": 2,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`API: LSK Transactions createTransactionObject creates a transaction object for unlockToken transaction 1`] = `
Object {
  "asset": Object {
    "unlockObjects": Array [
      Object {
        "amount": -1000n,
        "delegateAddress": Object {
          "data": Array [
            208,
            70,
            153,
            229,
            124,
            74,
            56,
            70,
            201,
            136,
            243,
            193,
            83,
            6,
            121,
            111,
            142,
            174,
            92,
            28,
          ],
          "type": "Buffer",
        },
        "unvoteHeight": undefined,
      },
      Object {
        "amount": 1000n,
        "delegateAddress": Object {
          "data": Array [
            183,
            120,
            252,
            149,
            185,
            240,
            116,
            7,
            230,
            64,
            157,
            115,
            200,
            175,
            25,
            25,
            217,
            3,
            80,
            2,
          ],
          "type": "Buffer",
        },
        "unvoteHeight": undefined,
      },
    ],
  },
  "assetID": 2,
  "fee": 1000000n,
  "moduleID": 5,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`API: LSK Transactions createTransactionObject creates a transaction object for vote transaction 1`] = `
Object {
  "asset": Object {
    "votes": Array [
      Object {
        "amount": 100n,
        "delegateAddress": Object {
          "data": Array [
            208,
            70,
            153,
            229,
            124,
            74,
            56,
            70,
            201,
            136,
            243,
            193,
            83,
            6,
            121,
            111,
            142,
            174,
            92,
            28,
          ],
          "type": "Buffer",
        },
      },
      Object {
        "amount": -100n,
        "delegateAddress": Object {
          "data": Array [
            183,
            120,
            252,
            149,
            185,
            240,
            116,
            7,
            230,
            64,
            157,
            115,
            200,
            175,
            25,
            25,
            217,
            3,
            80,
            2,
          ],
          "type": "Buffer",
        },
      },
    ],
  },
  "assetID": 1,
  "fee": 1000000n,
  "moduleID": 5,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`API: LSK Transactions elementTxToDesktopTx should a register delegate transaction with type signature of lisk service 1`] = `
Object {
  "asset": Object {
    "username": "super_delegate",
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "5:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions elementTxToDesktopTx should a transfer transaction with type signature of lisk service 1`] = `
Object {
  "asset": Object {
    "amount": "100000000",
    "data": "",
    "recipient": Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    },
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "2:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions elementTxToDesktopTx should a vote delegate transaction with type signature of lisk service 1`] = `
Object {
  "asset": Object {
    "votes": Array [
      Object {
        "amount": "100",
        "delegateAddress": "lskvj27a26",
      },
    ],
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "5:1",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions elementTxToDesktopTx should transform a reclaimLSK transaction 1`] = `
Object {
  "asset": Object {
    "amount": "100",
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "1000:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions elementTxToDesktopTx should transform a registerMultisignatureGroup transaction 1`] = `
Object {
  "asset": Object {
    "mandatoryKeys": Array [
      "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
      "86499879448d1b0215d59cbf078836e3d7d9d2782d56a2274a568761bff36f19",
    ],
    "numberOfSignatures": 2,
    "optionalKeys": Array [
      "35c6b25520fc868b56c83fed6e1c89bb350fb7994a5da0bcea7a4f621f948c7f",
    ],
  },
  "fee": "1000000",
  "id": "12",
  "moduleAssetId": "4:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions elementTxToDesktopTx should transform a unlockToken transaction 1`] = `
Object {
  "asset": Object {
    "unlockObjects": Array [
      Object {
        "amount": "10000000",
        "delegateAddress": "lskehj8am9afxdz8arztqajy52acnoubkzvmo9cjy",
        "unvoteHeight": 1000000,
      },
      Object {
        "amount": "-10000000",
        "delegateAddress": "lskkx46qy3wkwtfbh4gzbq57tx6dzfexu5yufwh5b",
        "unvoteHeight": 1000000,
      },
    ],
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "5:2",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions transactionToJSON should return the transaction as JSON 1`] = `"{\\"moduleID\\":2,\\"assetID\\":0,\\"senderPublicKey\\":\\"0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a\\",\\"nonce\\":\\"2n\\",\\"fee\\":\\"123123n\\",\\"signatures\\":[],\\"asset\\":{\\"recipientAddress\\":\\"b778fc95b9f07407e6409d73c8af1919d9035002\\",\\"amount\\":\\"10000n\\"}}"`;

exports[`API: LSK Transactions transformTransaction should a register delegate transaction with type signature of lisk service 1`] = `
Object {
  "asset": Object {
    "username": "super_delegate",
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "5:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions transformTransaction should a transfer transaction with type signature of lisk service 1`] = `
Object {
  "asset": Object {
    "amount": "100000000",
    "data": "",
    "recipient": Object {
      "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    },
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "2:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions transformTransaction should a vote delegate transaction with type signature of lisk service 1`] = `
Object {
  "asset": Object {
    "votes": Array [
      Object {
        "amount": "100",
        "delegateAddress": "lskvj27a26",
      },
    ],
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "5:1",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions transformTransaction should transform a reclaimLSK transaction 1`] = `
Object {
  "asset": Object {
    "amount": "100",
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "1000:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions transformTransaction should transform a registerMultisignatureGroup transaction 1`] = `
Object {
  "asset": Object {
    "mandatoryKeys": Array [
      "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
      "86499879448d1b0215d59cbf078836e3d7d9d2782d56a2274a568761bff36f19",
    ],
    "numberOfSignatures": 2,
    "optionalKeys": Array [
      "35c6b25520fc868b56c83fed6e1c89bb350fb7994a5da0bcea7a4f621f948c7f",
    ],
  },
  "fee": "1000000",
  "id": "12",
  "moduleAssetId": "4:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions transformTransaction should transform a unlockToken transaction 1`] = `
Object {
  "asset": Object {
    "unlockObjects": Array [
      Object {
        "amount": "10000000",
        "delegateAddress": "lskehj8am9afxdz8arztqajy52acnoubkzvmo9cjy",
        "unvoteHeight": 1000000,
      },
      Object {
        "amount": "-10000000",
        "delegateAddress": "lskkx46qy3wkwtfbh4gzbq57tx6dzfexu5yufwh5b",
        "unvoteHeight": 1000000,
      },
    ],
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "5:2",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;
